audio/out/ao_wasapi.c

```diff
@@ -190,7 +190,13 @@ static void set_state_and_wakeup_thread(struct ao *ao,
 
 static void thread_process_dispatch(void *ptr)
 {
-    set_state_and_wakeup_thread(ptr, WASAPI_THREAD_DISPATCH);
+    struct wasapi_state *state = ((struct ao *)ptr)->priv;
+    uint32_t seq;
+
+    do {
+        seq = atomic_fetch_add(&state->dispatch_seq, 1);
+    } while (seq < WASAPI_THREAD_DISPATCH_BASE);
```

Comment:
What does it do? You just loop through all values.

Reply:
To give each thread control call a different state value (>= `WASAPI_THREAD_DISPATCH_BASE`).

Reply:
Ok, I see now. There is compare exchange and if thread state changed we process this state, else we go back to feed.

But is it really working as intended? It would only fix the race between consecutive dispatch wakeups. But what if you get for example any other state like `WASAPI_THREAD_RESET` it would skip any dispatch "state" or even any other state. 

So only last wakeup state is processed either way. Incrementing counter make it so dispatch wakeups are distinct, but they still can be overwrite by another state.

Wouldn't it be better actually to not overwrite thread_state with dispatch? And always get process the dispatch queue if there is anything there?

Reply:
> So only last wakeup state is processed either way.

True. But reset/resume/control requests are sent from the same thread (for now at least) and AO control is a synchronous call, so this shouldn't be a problem.

> Wouldn't it be better actually to not overwrite thread_state with dispatch? And always process the dispatch queue if there is anything there?

This effectively reverts 6863eefc3d1d683a08f56aeaca0b4706672f2447. The bottom line is we cannot call `thread_feed` on any manual wakeup or there will be guaranteed glitches and desync.

Reply:
I kind of dislike this `thread_state`  and `dispatch_seq` business. I understand the idea, but it feels little bit off to me. Maybe there is  no problem in practice, but how about we base it on events instead?

How about we do `WaitForMultipleObjects` and have dedicated event for dispatch work. Then we just check which event woken us up and do dispatch or feed.

Reply:
> How about we do `WaitForMultipleObjects` and have dedicated event for dispatch work.

I actually went for this at first but there seems to have been trouble when using `(Msg)WaitForMultipleObjects` (https://github.com/mpv-player/mpv/pull/2872):

> Instead the under-runs were the result of MsgWaitForMultipleObjects not returning at consistent intervals. It was indeed discovered that using a simple WaitForSingleObject on the thread wakeup event eliminated these under-runs.

But well that was a decade ago. Maybe things have changed since that.

I'll give it a try later :)

Reply:
Note that we should not use `MsgWaitForMultipleObjects` on this thread, but directly `WaitForMultipleObjects`. Not sure if this makes practical difference, but just in case.